### ISSUE TITLE
Hide opportunistic locking within Ganeti

### ISSUE THREAD

**Issue by Ganeti-Issues-Migrator** [ganeti/ganeti#911]
Originally reported of Google Code with ID 855.

```
Opportunistic locking is a feature of instance creation that allows the iallocator to grab only the available node locks for determining instance placement.

The problem is that this operation can fail if not enough nodes are available, and retries are currently handled by the user. With the instance moves already complicated due to security considerations, anyone not wanting to use the move-instance tool is bound to have fun getting cross-cluster instance moves to work.

This would be better handled by modifying the locking mechanisms of Ganeti to allow for certain constraints: e.g. grab as many locks as possible, but no less than two.


```
Originally added on 2014-06-13 15:48:47 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Also not that "any two node locks" is not enough: you might be unlucky and get nodes with not enough resources, while other nodes still have enough resources. So after getting insufficient resources on the opportunistically acquired locks, we should verify if the cluster as a whole has enough resources. If not, we can fail, but if it does have enough resources we retry asking for a set containing at least two of the suitable nodes.
```
Originally added on 2014-06-13 17:11:44 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Perhaps (while fixing the bug) it'd be feasible to add a QA test for this using the new parallel testing QA tools.
```
Originally added on 2014-06-19 20:29:20 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
commit f3f1fc574671d41fefb05661b01d4ff93312eef7
Author: Klaus Aehlig <aehlig@google.com>
Date:   Tue Jul 8 17:29:41 2014 +0200

    Honor requests for minimal lock count
    
    Change lock allocation in mcpu to honor the required minimum
    of locks to allocate in opportunistic lock allocation.
    
    Signed-off-by: Klaus Aehlig <aehlig@google.com>
    Reviewed-by: Hrvoje Ribicic <riba@google.com>

commit ed94242fdd9947815c84de8b0727aa63630c8f34
Author: Klaus Aehlig <aehlig@google.com>
Date:   Tue Jul 8 17:26:30 2014 +0200

    Make instance creation ask for 2 nodes, if DRBD is used
    
    In opportunistic locking, we always ask for as many locks
    as possible to later chose what fits best. However, we
    may also ask for a minimum as otherwise it cannot work
    for principle reasons. For DRBD, this minimum should
    be two, rather than the default one.
    
    Signed-off-by: Klaus Aehlig <aehlig@google.com>
    Reviewed-by: Hrvoje Ribicic <riba@google.com>

commit 24ef93b37ba2411e86f06e045d523a8769372335
Author: Klaus Aehlig <aehlig@google.com>
Date:   Tue Jul 8 17:03:48 2014 +0200

    Allow LUs to specify a number of opportunistic locks
    
    LUs may decide to acquire locks in an opportunistic way. In this
    case, also allow them to specify the number of locks the which to
    acquire at least.
    
    Signed-off-by: Klaus Aehlig <aehlig@google.com>
    Reviewed-by: Hrvoje Ribicic <riba@google.com>

commit 280f37bccdc676aec4e6e5aaf7e321632b12756b
Author: Klaus Aehlig <aehlig@google.com>
Date:   Tue Jul 8 16:53:05 2014 +0200

    Export guardedOpportunisticLockUnion through WConfD
    
    ...so that jobs can wait till they get a number of locks
    that might work at least in principle.
    
    Signed-off-by: Klaus Aehlig <aehlig@google.com>
    Reviewed-by: Hrvoje Ribicic <riba@google.com>

commit 96245004980fdf9b8f2257a52ada81b08404b75b
Author: Klaus Aehlig <aehlig@google.com>
Date:   Tue Jul 8 14:44:16 2014 +0200

    Add a guarded version of opportunistic locking
    
    ...allowing to request at least a certain number of
    the requests to succeed and otherwise not touch the
    locking state.
    
    Signed-off-by: Klaus Aehlig <aehlig@google.com>
    Reviewed-by: Hrvoje Ribicic <riba@google.com>

```
Originally added on 2014-07-10 11:25:07 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
commit 37d6f7f5f7ebfc3f40b70c28930a99045aa1cd01
Author: Klaus Aehlig <aehlig@google.com>
Date:   Wed Jul 9 17:48:43 2014 +0200

    Test parallel instance creation with --opportunistic-locking
    
    With the recent changes, opportunistic locking should not
    require retries any more on clusters that have enough capacity.
    So it is fair to add the --opportunistic-locking option the
    parallel instance creation tests. In this way, we
    - verify that the absence of the need to retry is indeed true and
    - get a realistic performance measurement for truly parallel
      instance creation.
    
    Signed-off-by: Klaus Aehlig <aehlig@google.com>
    Reviewed-by: Hrvoje Ribicic <riba@google.com>

commit a305cb96e24303414ed4512932be8208fc5fdcca
Author: Klaus Aehlig <aehlig@google.com>
Date:   Wed Jul 9 17:36:50 2014 +0200

    Clarify setup for parallel job execution performance test
    
    Explicitly state, that parallelism enhancing options are to be used
    in parallel performance tests, if it is best practice to use them on
    the respective versions. A prerequisite for those options is that
    using them does not require, even under heavy load but assuming enough
    capacity in the cluster, any jobs to be retried.
    
    Signed-off-by: Klaus Aehlig <aehlig@google.com>
    Reviewed-by: Hrvoje Ribicic <riba@google.com>

```
Originally added on 2014-07-10 11:42:00 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Note that this has been cherry-picked into 2.12:

[6f2fb54] Test parallel instance creation with --opportunistic-locking
[a5bccf5] Clarify setup for parallel job execution performance test
[4463621] Honor requests for minimal lock count
[85b2005] Make instance creation ask for 2 nodes, if DRBD is used
[45c4735] Allow LUs to specify a number of opportunistic locks
[a001c5c] Export guardedOpportunisticLockUnion through WConfD
[d434a2c] Add a guarded version of opportunistic locking

```
Originally added on 2014-08-29 10:31:18 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
-- Empty comment --
```
Originally added on 2014-09-10 12:39:50 +0000 UTC.

Now blocked by: 940 (now #996);

**Comment by Ganeti-Issues-Migrator** [comment]
```
commit 3b4296047a477a99dba84aed50295a16a9e34cf3
Author: Klaus Aehlig <aehlig@google.com>
Date:   Mon Sep 22 16:38:34 2014 +0200

    Explicitly log failed opportunistic locking
    
    Add an additional log entry if the IAllocator failed to find
    a solution in opportunistic locking. Otherwise the IAllocator
    message would get lost on internal retries.
    
    Signed-off-by: Klaus Aehlig <aehlig@google.com>
    Reviewed-by: Petr Pudlak <pudlak@google.com>
commit 70a71730aa2569428d587ef45fefde80b8d8cf2c
Author: Klaus Aehlig <aehlig@google.com>
Date:   Mon Sep 22 14:33:24 2014 +0200

    Allow internal retries in LUInstanceCreate
    
    ...provided we failed to obtain enough resources with opportunistic
    locking. In this case, we fall back to full (non-opportunistic) lock
    allocation.
    
    Signed-off-by: Klaus Aehlig <aehlig@google.com>
    Reviewed-by: Petr Pudlak <pudlak@google.com>

```
Originally added on 2014-09-23 14:02:51 +0000 UTC.

Changed State: Fixed

**Comment by Ganeti-Issues-Migrator** [comment]
```
Released as part of 2.14.0
```
Originally added on 2015-06-02 10:37:02 +0000 UTC.

Changed State: Released
